server/app/services/applicant/ApplicantService.java

```diff
@@ -837,7 +848,7 @@ public CompletionStage<ApplicationPrograms> relevantProgramsForApplicant(long ap
                   applications.stream()
                       .map(application -> application.getProgram().getProgramDefinition())
                       .collect(Collectors.toList());
-              programDefinitionsList.addAll(activeProgramDefinitions);
+              programDefinitionsList.addAll(activeProgramDefinitionsFuture.join());
```

Comment:
There's no telling if this particular future has completed yet or not so this may block the calling thread. If you make L844 `CompletableFuture.allOf(applicationsFuture, activeProgramDefinitionsFuture, ...) then calling `join` here is guaranteed not to block.

Reply:
I need to pass the output of `applicationsFuture` into the `thenComposeAsync`, does that work with `allOf`?